### PR TITLE
Prison visits virtual from json

### PR DESCRIPTION
**Pull request by neilblair**
Alter prison visit booking handler to retrieve all visit slots, face-to-face slots and virtual visit slots, from data posted from PRISM.  (Previously face-to-face slots came from PRISM data and virtual slots were notionally created from config)

Add a submission serial number to emailed form submission to help identify when a submission has not been received.

### CODE REVIEW

File: web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php [hunk 1]
```diff
@@ -800,28 +763,54 @@ private function getAvailableSlots() {
...
-        $updated_slots[] = $slot_date_adjusted;
+    }
+    else {
+
+      // No data in cache, so try file instead. Every time the
+      // external service posts data to prison visits api controller,
+      // data is stored in cache and written to file.
+
+      $file_uri = 'private://nidirect_webforms/prison_visit_slots_data.json';
+      $file_contents = @file_get_contents($file_uri);
```
Comment (omahm) [inline web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php:802]: Using @ is not recommended. I'd use file_exists to check the file and then read the contents.

File: web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php [hunk 2]
```diff
@@ -800,28 +763,54 @@ private function getAvailableSlots() {
...
+
+      if (!empty($file_contents)) {
+        $data = json_decode($file_contents, TRUE);
```
Comment (omahm) [inline web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php:805]: PHP 8 introduces the json_validate method that can be used to check if the file contents are valid JSON and and errors can be retrieved with json_last_error()
https://www.php.net/manual/en/function.json-validate.php

**Review by omahm — CHANGES_REQUESTED** [review]
Looks good, just a couple of defensive coding suggestions.

File: web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php (hunk 2 above)
Comment (neilblair) [reply web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php:805]: The JSON is validated on the way in by the [PrisonVisitBookingJsonApiController.php](https://github.com/dof-dss/nidirect-drupal/blob/prison-visits-virtual-from-json/web/modules/custom/nidirect_webforms/src/Controller/PrisonVisitBookingJsonApiController.php) route controller around line 188.  

So the data is only saved to cache and to the file if it is valid.  Is that enough or should it be validated when retrieved as well?

Comment (omahm) [reply web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php:805]: Looks good 👍

Comment (neilblair) [reply web/modules/custom/nidirect_webforms/src/Plugin/WebformHandler/PrisonVisitBookingHandler.php:805]: I suppose there might be a possibility the file becomes corrupted ...?

**Review by omahm — APPROVED** [review]

**Pull request merged**